Fix vec iter zst alignment

DrAsu33 opened the pull request:
<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

Closes rust-lang/rust#148682

reddevilmidzy commented on tests/ui/issues/issue-148682.rs:
It looks like the file name needs to be changed.

https://rustc-dev-guide.rust-lang.org/tests/best-practices.html#test-naming

reddevilmidzy replied:
Instead of using the issues directory, please refer to [this](https://github.com/rust-lang/rust/blob/main/tests/ui/README.md) and choose a more appropriate location.

DrAsu33 commented:
@reddevilmidzy Thanks! Updated as requested.

reddevilmidzy commented on tests/ui/issues/issue-zst-nthback-148682.rs:
```diff
@@ -0,0 +1,9 @@
+// @ check-pass
+//test Intolter::nth_back does not cause UB for ZSTs with high alignment
```
Please remove leading spaces from commands, and add a space before comments.

```suggestion
//@ check-pass
// test Intolter::nth_back does not cause UB for ZSTs with high alignment
```

reddevilmidzy reviewed (commented):
and please squash

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/149272/42ec52babac2cbf2bb2b9d794f980cbcb3ebe413..96102b7b41fb28dd90207ec58459b6675475a47d)*

DrAsu33 committed:
Fix(alloc): Correctly handle ZST alignment for IntoIter::nth_back

This commit consolidates all changes, including the core logic fix for IntoIter::nth_back and the addition of the Miri regression test in `library/alloctests/tests/vec.rs`, to prevent Undefined Behavior (UB) when dealing with highly-aligned Zero-Sized Types.

workingjubilee commented on tests/ui/iterators/ZST-nthback.rs:
```diff
@@ -0,0 +1,10 @@
+//@ run-pass
```
This still will not be a test we run under miri, I believe.

We do run the standard library's tests under miri, however.

RalfJung commented on src/tools/miri/tests/fail/unaligned_pointers/ZST-nthback.rs:
```diff
@@ -0,0 +1,27 @@
+//@ compile-flags: -Zmiri-symbolic-alignment-check
+//@ run-pass
+
+use std::ptr;
```
What are you importing `std::ptr` for?

RalfJung reviewed (commented):
Some comments for the Miri test. :)
I did not look at the libs part.

FWIW, we do run the standard library test suite in Miri. So maybe it would make more sense to add this there? We already have some tests in the standard library test suite that are primarily meant for Miri.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/149272/467250ddb25a93315a8dee02dd6cc1e398be46ff..d7b95bcdee562e19b4504c60503a5434480b5963)*

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/19781771055/job/56682980165) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/56682980165) [(plain)](https://github.com/rust-lang/rust/commit/2849232e8c71b2b017917afac588f9f27ee47d7a/checks/56682980165/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Diff in /checkout/library/alloctests/tests/vec.rs:2718:
     unsafe { new.read() };
 }
 
-#[test] 
-#[cfg(miri)] 
+#[test]
+#[cfg(miri)]
 fn zst_vec_iter_nth_back_regression() {
     // Regression test for Undefined Behavior (UB) caused by IntoIter::nth_back (#148682)
     // when dealing with high-aligned Zero-Sized Types (ZSTs).
Diff in /checkout/library/alloctests/tests/vec.rs:2726:
-    // This test ensures the fix (using ptr::NonNull::dangling() internally) 
-    // is correctly verified when Miri runs it without reporting any UB.  
+    // This test ensures the fix (using ptr::NonNull::dangling() internally)
+    // is correctly verified when Miri runs it without reporting any UB.
```

</details>

RalfJung commented on library/alloctests/tests/vec.rs:
```diff
@@ -2718,21 +2718,16 @@ fn vec_null_ptr_roundtrip() {
+// Regression test for Undefined Behavior (UB) caused by IntoIter::nth_back (#148682)
+// when dealing with high-aligned Zero-Sized Types (ZSTs).
 #[test] 
 #[cfg(miri)] 
```
Why do you only enable this test in Miri?

DrAsu33 commented:
Builds are green and I've addressed all feedback (moved test to library/alloc/tests/vec.rs).

@rustbot ready

Mark-Simulacrum reviewed (changes requested):
r=me with extra tests added

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/149272/3c5c55864f8f441054730037bc9f9901f9151398..89b4b309b84f45901dd42a65d250549c3070e22f)*
